Platform/Plugins/com.tle.platform.common/src/com/tle/common/util/ExecUtils.java

```
@@ -52,6 +55,71 @@
+      try {
+        if (Integer.parseInt(args[0]) > 0) {
+          // get child process PIDs as string
+          Process getChildPid = Runtime.getRuntime().exec("pgrep -P " + args[0]);
+          getChildPid.waitFor();
+          StringBuilder childPid = new StringBuilder();
+          if (getChildPid.exitValue() == 0) {
+            InputStream procIn = getChildPid.getInputStream();
+            int c = 0;
+            while ((c = procIn.read()) != -1) {
+              childPid.append((char) c);
+            }
+          }
+
+          // Kill child process(es)
+          Process childKillProc = Runtime.getRuntime().exec("kill -9 " + childPid);
+          childKillProc.waitFor();
+          if (childKillProc.exitValue() == 0) {
+            LOGGER.debug("Child processes terminated: " + childPid);
+          }
+
+          // kill process itself
+          Process proc = Runtime.getRuntime().exec("kill -9 " + args[0]);
+          int exitVal = proc.waitFor();
+          LOGGER.debug("Exit value: " + exitVal); // often 1 under Ubuntu 12.10
+
+          proc.destroy();
+          childKillProc.destroy();
+          getChildPid.destroy();
+        }
+      } catch (Exception e) {
+        LOGGER.warn("Process Kill failed. Process may be left hanging");
+      }
+    }
+  }
+
+  /**
+   * Gets the PID of a given process.
+   *
+   * @param p The Process of which to get the PID.
+   * @return An Optional long. If not on Linux, or if the PID declared field is not available, the
+   *     value will be empty.
+   */
+  public static synchronized Optional<Long> getPidOfProcess(Process p) {
+    Optional<Long> pid = Optional.empty();
+
+    try {
+      if (p.getClass().getName().equals("java.lang.UNIXProcess")) {
+        Field f = p.getClass().getDeclaredField("pid");
+        f.setAccessible(true);
+        pid = Optional.of(f.getLong(p));
+        f.setAccessible(false);
+      }
+    } catch (Exception e) {
+      return pid;
+    }
+    return pid;
```

Seems awkward to have the two `return` statements right next to each other. Would a `finally` block perhaps be a better fit?

---

I tried that, but apparently the inspector has a problem with having a return within a finally block. 

>'return' inside 'finally' block 
 Inspection info: Reports return statements inside of finally blocks. While occasionally intended, such return statements may mask exceptions thrown, and tremendously complicate debugging.

---

I see it's now refactored to simply use the default return with `pid`. Much better.